File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        vector<string>res;

        dfsserialize(res,root);

        return join(res,",");
        
    }

    void dfsserialize(vector<string>&res,TreeNode*node)
    {
        if(!node) 
        {
            res.push_back("N");
            return;
        }

        res.push_back(to_string(node->val));
        dfsserialize(res,node->left);
        dfsserialize(res,node->right);

    }

    string join(const vector<string>& v, const string& delim) {
    ostringstream s;
    for (size_t idx = 0; idx < v.size(); ++idx) {
        if (idx) s << delim;     // add delimiter before non-first elements
        s << v[idx];
    }
    return s.str();              // safe even when v is empty
}

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        vector<string>vals=split(data,',');
        int i=0;
       return dfsdeserialize(i,vals);


        
    }

    TreeNode*dfsdeserialize(int&i, vector<string>&vals)
    {
        if(vals[i]=="N")
        {
            i++;
            return NULL;
        }

        TreeNode*root=new TreeNode(stoi(vals[i]));
        i++;
        root->left=dfsdeserialize(i,vals);
        root->right=dfsdeserialize(i,vals);

        return root;
    }

    vector<string> split(const string &s, char delim) {
        vector<string>elems;
        stringstream ss(s);

        string item;
        while(getline(ss,item,delim))
        {
            elems.push_back(item);
        }
        return elems;
    }

    



    
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """


        ans=[]

        def dfs(node):

            if not node:
                ans.append('#')
                return 

            ans.append(node.val)
            dfs(node.left)
            dfs(node.right)




        dfs(root)

        return ','.join(map(str,ans))
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """


        idx=[0]

        tokens=data.split(',')

        def dfs():
            if tokens[idx[0]]=='#':
                idx[0]+=1
                return None
            node=TreeNode(tokens[idx[0]])
            idx[0]+=1
            
            node.left=dfs()
            node.right=dfs()



            return node


        return dfs()

        

        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))